### DISCUSSION TITLE
Outdated page with on-demand revalidation and client-side navigation

### DISCUSSION THREAD

**Discussion by sspenst** [vercel/next.js#35128] [Help] [answered]
I'm trying out the new on-demand revalidation and I've been running into an issue with my use case:

I have a leaderboard page that I'm updating with unstable_revalidate and I want to be able to navigate to the leaderboard with client-side navigation and see the updated leaderboard.

Currently when navigating to the leaderboard with a `Link` you still see the original leaderboard from when the page was first loaded. A refresh is needed to get latest leaderboard.

My current workaround is to replace the leaderboard `Link` tag with an `a` tag, but I'm wondering if there's a more official way to do this so I can keep the client-side navigation. I've tried setting `prefetch={false}` in the `Link` but it seems this isn't supported for pages using ISG...

Let me know if I'm missing anything!

**Comment by sspenst** [comment]
This is a probably a separate issue, but it seems revalidation will randomly take >10s to complete simple DB queries (eg finding 6 MongoDB documents).

I've decided to stop using the revalidate feature for now and instead use getServerSideProps for my leaderboard page. It takes a bit longer to load the page but at least it's consistent.

**Reply by icyJoseph** [reply]
I don't think revalidate is necessary to do this, nor do I think it is the use case for it, or least it shouldn't be used alone.

There's no guarantee on the duration and result of a revalidate option, and trying to build a UI that relies on the revalidate process being done "quick", does feel problematic. For example, the CPU might be busy, and hence the revalidate takes longer, or there's no ongoing connection to the DB, so opening a new one adds overhead, etc, these are all just things that one must think of when it comes to asynchronous work.

It does feel like `getServerSideProps` and a client side data fetching solution should do the trick for you (like SWR or React-Query), and perhaps you need an additional `api/route`, or refer to this trick by Josh Comeau, https://www.joshwcomeau.com/nextjs/refreshing-server-side-props/.

A client side data fetching solution, with revalidate (GetStaticProps), could also work, helping the UI follow the server state, while also statically rendering updated versions from the revalidate process. 

However, as said before, there's no guarantee on how long it will take, so even if the UI agrees with the server data, if the client does a page refresh, you might flash the user with old data, unless you don't pre-render the dynamic data, only the App Shell, and always get all the data from the client, using an `api/route`, at this point then revalidate is no longer necessary.

**Reply by sspenst** [reply]
Thanks for the response! It seems like my main issue was relying too heavily on `unstable_revalidate`. I'm now using your suggestion of client-side data fetching with revalidate and it seems to be much more stable. For reference my setup is now:

- `getStaticProps` to generate the leaderboard page in advance
- Call `unstable_revalidate` when the leaderboard needs to be updated, but don't await the response (allowing for potential failures when revalidating)
- `useSWR` on the leaderboard page so that you eventually end up with the latest data
  - [This page](https://swr.vercel.app/docs/with-nextjs) was helpful in getting SSG set up with SWR

**Comment by sspenst** [comment, marked as answer]
My solution was to change my approach to the following:

- `getStaticProps` to generate the leaderboard page in advance
- Call `unstable_revalidate` when the leaderboard needs to be updated, but don't await the response (allowing for potential failures when revalidating)
- `useSWR` on the leaderboard page so that you eventually end up with the latest data
  - [This page](https://swr.vercel.app/docs/with-nextjs) was helpful in getting SSG set up with SWR
  - Note this also allows using `Link` for the leaderboard page since the data will be retrieved with SWR
